libsolidity/formal/Z3CHCSmtLib2Interface.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/formal/Z3CHCSmtLib2Interface.h>
+
+#include <libsolidity/interface/UniversalCallback.h>
+
+#include <libsmtutil/SMTLib2Parser.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <stack>
+
+using namespace solidity::frontend::smt;
+using namespace solidity::smtutil;
+
+Z3CHCSmtLib2Interface::Z3CHCSmtLib2Interface(
+	frontend::ReadCallback::Callback _smtCallback,
+	std::optional<unsigned int> _queryTimeout,
+	bool _computeInvariants
+): CHCSmtLib2Interface({}, std::move(_smtCallback), _queryTimeout), m_computeInvariants(_computeInvariants)
+{
+}
+
+void Z3CHCSmtLib2Interface::setupSmtCallback(bool _enablePreprocessing)
+{
+	if (auto* universalCallback = m_smtCallback.target<frontend::UniversalCallback>())
+		universalCallback->smtCommand().setZ3(m_queryTimeout, _enablePreprocessing, m_computeInvariants);
+}
+
+std::tuple<CheckResult, Expression, CHCSolverInterface::CexGraph> Z3CHCSmtLib2Interface::query(smtutil::Expression const& _block)
+{
+	setupSmtCallback(true);
+	std::string query = dumpQuery(_block);
+	std::string response = querySolver(query);
+	if (boost::starts_with(response, "unsat"))
+	{
+		// Repeat the query with preprocessing disabled, to get the full proof
+		setupSmtCallback(false);
+		query = "(set-option :produce-proofs true)" + query + "\n(get-proof)";
+		response = querySolver(query);
+		setupSmtCallback(true);
+		if (!boost::starts_with(response, "unsat"))
+			return {CheckResult::SATISFIABLE, Expression(true), {}};
+		return {CheckResult::SATISFIABLE, Expression(true), graphFromZ3Answer(response)};
```

Comment:
```suggestion
		return {CheckResult::UNSATISFIABLE, Expression(true), graphFromZ3Answer(response)};
```

Shouldn't this be unsatisfiable?

Reply:
This is a bit confusing but let me explain:
When you use Z3 as a library, it uses the original Datalog semantics where SAT means UNSAFE and UNSAT means SAFE.
SMTChecker followed this convention.
However, when you use SMT-LIB format with `HORN` logic, the semantics is flipped. SAT means SAFE and UNSAT means UNSAFE.
That's why we need to flip the answer from the solver when we propagate it.
See also `CHCSmtLib2Interface::query` where the same flipping is implemented as well.

Reply:
This is indeed confusing, and just a little bit insane, so a comment outlying this would be very much appreciated :)

Reply:
It is very confusing if you are not familiar with CHCs.
The thing is that it flips the question you are asking.
In the usual SMT-based model checking you are asking the solver "Is there a counterexample path?"
With Horn solvers, you are asking a different question, "Are there safe inductive invariants?"
So the meaning of the YES/NO answer is flipped.

I'll add a comment in the code.

Reply:
Comment added here and also to `CHCSmtLib2Interface`